### ISSUE TITLE
Support for Arc browser

### ISSUE THREAD

**Issue by tom-sherman** [fwdcloudsec/granted#357]
https://arc.net/

Like firefox, it has the concept of containers that it calls "profiles". Would be great if `assume -c` could open Arc.

**Comment by jayprice-wd** [comment]
Agreed. Another vote for this. As it's chrome based, it should be a similar implementation.

**Comment by CodyDunlap** [comment]
Also looking for this. At the moment, I'm using the `-u` option to provide a link that can be clicked to open the default browser.

**Comment by guillaume-mindflow** [comment]
Would love to see that as well!

**Comment by tom-sherman** [comment]
This still hasn't replaced Firefox for me. Unfortunately it doesn't allow you to open multiple AWS profiles at once, when you open the second one it'll say you're already logged in.

**Comment by shwethaumashanker** [comment]
Hey @tom-sherman! Thanks for the update. I've reopened the issue, and we'll take a closer look and keep you posted. 😊

**Comment by Sandy-SecOps** [comment]
Would love to see this added - I love Arc browser, and love granted, but stuck with FF untill support has landed. Any update?

**Comment by chrnorm** [comment]
An update here - currently it appears that Arc doesn't have the necessary CLI flags for us to programatically open a tab in a particular profile. Some extra details I found:

https://www.reddit.com/r/ArcBrowser/comments/19bl4gy/support_choosing_profile_via_commandline_flag/

https://github.com/Browsers-software/browsers/issues/94

I've just raised a feature request with the Arc team for this - I'll add any replies I get from their team into this issue.

**Comment by chrnorm** [comment]
@ScottGuymer it looks like at the moment you need to use AppleScript with Arc (see [this Reddit thread](https://www.reddit.com/r/ArcBrowser/comments/120d4vl/command_line_to_open_little_arc/)), could you try something like the following and see if it works?

```bash
# create the file ~/.granted/launch-arc.sh and make it executable, with these contents

#!/usr/bin/env bash

osascript -e "tell application \"Arc\"
  make new tab with properties {URL:\"$1\"}
  activate
end tell"
```

```bash
# ~/.granted/config
[AWSConsoleBrowserLaunchTemplate]
Command = "/full/path/to/.granted/launch-arc.sh {{.URL}}"
```

**Comment by ScottGuymer** [comment]
It doesn't seem to work for me.

nothing seemed to change and it still tried to open another instance of Arc

**Comment by gcaride** [comment]
> At the moment I don't need any profile support in the browser. I have configured arc but get this error when trying to login
> 
> ![image](https://private-user-images.githubusercontent.com/956825/369319040-e9dced1f-54ce-491c-aa8e-39db7bdd6dac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyOTg5NzMsIm5iZiI6MTczMDI5ODY3MywicGF0aCI6Ii85NTY4MjUvMzY5MzE5MDQwLWU5ZGNlZDFmLTU0Y2UtNDkxYy1hYThlLTM5ZGI3YmRkNmRhYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAzMFQxNDMxMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTBjY2Q3OGU2Mjg3YzQyODI2YjVkNWFhNWRiN2JlNGQ2NzkwOGM3YzRjMjU1MGNhZjFiYjUyYWI4Zjk5MTkyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VdcNWA9Lqc_8W_koj_C6XyEsIwwkdnFzFYJdxJu9lQM)
> 
> I would expect it to just open a browser session

I had that same issue too. 

I was able to make it work with this configuration in the `.granted/config` 

```
DefaultBrowser = "STDOUT"
CustomBrowserPath = ""
CustomSSOBrowserPath = ""
Ordering = ""
ExportCredentialSuffix = ""

[Keyring]
  Backend = "keychain"
```

**Comment by mmathys** [comment]
I was able to make it work by setting the browser to `Custom` when running `granted browser set-sso`.

In `~/.granted/config`, the entry should be blank:

```
...
CustomSSOBrowserPath = ""
...
```

Was able to use assume with Arc without any problems.

**Comment by jvanbrunt** [comment]
> I was able to make it work by **setting the browser to `Custom`** when running `granted browser set-sso`.
> 
> In `~/.granted/config`, the entry should be blank after running the command:
> 
> ```
> ...
> CustomSSOBrowserPath = ""
> ...
> ```
> 
> Was able to use assume with Arc without any problems.

This worked for me!

**Comment by meyerjrr** [comment]
Hey All! I've taken a look into this and found there is a _potential_ work-around. I'll preface this by saying that it's still not quite the same as the UX Granted has in Firefox where you can achieve different profiles per tab, that is currently not possible as per @chrnorm comment above

Found this online that explains how it works in Arc:
> You can't have more than one profile in a space. Each space can be assigned a different profile. You can switch profiles easily by switching spaces. If you want a different profile for each client, you would need to create a space for each client and then assign a corresponding profile to each workspace.

The workaround is to have Granted create new spaces + profiles for new AWS profiles as containers which would achieve siloed effect desired. 

This would require programatically adding new spaces and profiles to the Arc config.

All config is handled locally inside of `StorableSidebar.json` and `StorableLinkRouting.json` inside the application settings on OSX. I haven't delved into other OS's yet to see if this is achievable, I think its just windows that has Arc browser support as well now.

> How does it know what space + profile to open up into?

Arc has a config setting called "Air Traffic Control" which allows you to route which space to open up into based on a URL pattern. Here I used the profile that is included in the federated URL that opens up the console. ie `Profile=jack-sandbox1` is a query param in the URL.

![Image](https://github.com/user-attachments/assets/cff5ceba-dbe7-4d4b-ac62-a1bf6d56a8c3)

Similarly to the spaces and profiles config these routes are configurable programatically via config.

I have tested this by hardcoding the values in the config to confirm this will work for opening up into different profiles successfully, and have made a start on writing these values into the config automatically now.
I'll have a PR up with a POC soon, but just checking in here to see if the demand is still here / the proposed UX is something worth using for those Arc users out there!


There is also the addition of AWS adding partial support for multiple profiles open at once which may have deprioritised this issue:
https://docs.aws.amazon.com/awsconsolehelpdocs/latest/gsg/multisession.html

**Comment by meyerjrr** [comment]
Going to close this as Arc is now an unsupported browser by the Browser company, and does not look like we will get a solve for the above. Arc can still be used with granted (without the profiles) if needed
